pkg/scripts/install.go

```diff
@@ -13,177 +15,234 @@ import (
+	failedScripts := []string{}
+	var err error
+	for _, r := range result {
+		if r.Error != nil {
+			failedScripts = append(failedScripts, r.Script)
+		}
+	}
+	if len(failedScripts) > 0 {
+		err = errors.NewInstallRequiredError(failedScripts, docker.ScriptsURLLabel)
+	}
+	return result, err
+}
 
-		// and possible locations
-		for _, location := range locationsOrder {
-			locationResults, ok := locationsResultsMap[location]
-			if !ok || locationResults == nil {
-				continue
-			}
-			downloadResult, ok := locationResults[script]
-			if !ok {
-				continue
-			}
-			result.URL = downloadResult.location
-
-			// if location results are erroneous we store error in result object
-			// and continue searching other locations
-			if downloadResult.err != nil {
-				// one exception is when error contains information about scripts being inside the image
-				if e, ok := downloadResult.err.(errors.Error); ok && e.ErrorCode == errors.ScriptsInsideImageError {
-					// in which case update result object and break further searching
-					result.Error = nil
-					result.Downloaded = false
-					result.Installed = true
-					break
+// InstallOptional downloads and installs a set of scripts into dstDir, the result is a
+// map of scripts with detailed information about each of the scripts install process
+func (i *DefaultScriptSourceManager) InstallOptional(scripts []string, dstDir string) []api.InstallResult {
+	result := []api.InstallResult{}
+	for _, script := range scripts {
+		installed := false
+		for e := i.sources.Front(); e != nil; e = e.Next() {
+			detected := false
+			h := e.Value.(ScriptHandler)
+			h.SetDestinationDir(dstDir)
+			if r := h.Get(script); r != nil {
+				if err := h.Install(r); err != nil {
+					glog.Errorf("script %q found by the %s, but failed to install: %v", script, handlerName(h), err)
 				} else {
-					result.Error = downloadResult.err
-					continue
+					result = append(result, *r)
+					installed = true
+					detected = true
+					glog.Infof("Using %q installed from %q", script, r.URL)
 				}
 			}
-
-			// if there was no error
-			src := filepath.Join(dstDir, location, script)
-			dst := filepath.Join(dstDir, api.UploadScripts, script)
-			// move script to upload directory
-			if err := i.fs.Rename(src, dst); err != nil {
-				result.Error = err
-				continue
-			}
-			// set appropriate permissions
-			if err := i.fs.Chmod(dst, 0755); err != nil {
-				result.Error = err
-				continue
+			if detected {
+				break
 			}
-			// and finally update result object
-			result.Error = nil
-			result.Downloaded = true
-			result.Installed = true
-			break
 		}
-		resultList[idx] = result
+		if !installed {
+			result = append(result, api.InstallResult{
+				Script: script,
+				Error:  fmt.Errorf("script %q not installed", script),
+			})
+		}
 	}
+	return result
+}
 
-	return resultList
+// handlerName returns the human friendly name of the script handler.
+func handlerName(s ScriptHandler) string {
```

Comment:
Instead of type switch here, which will have to be updated whenever we add a handler I'd rather see this as a handler interface method, then everything you call against that interface is simpler to read.